core/src/main/java/io/kestra/core/server/ServiceLivenessManager.java

```diff
@@ -0,0 +1,380 @@
+                final ServiceInstance instance = localServiceState.instance();
+                final Service service = localServiceState.service();
+
+                // Execute the before hook.
+                if (!beforeScheduledStateUpdate(now, service, instance)) {
+                    return;
+                }
+
+                // Execute state update for current service (i.e., heartbeat).
+                updateServiceInstanceState(now, service, instance.state(), onStateTransitionFailureCallback);
+                log.trace("[Service id={}, type={}, hostname='{}'] Completed scheduled state update: '{}' ({}ms).",
+                    instance.id(),
+                    instance.type(),
+                    instance.server().hostname(),
+                    instance.state(),
+                    System.currentTimeMillis() - start
+                );
+            });
+    }
+
+    /**
+     * This method can be overridden to provide a hook before executing a scheduled service state update.
+     *
+     * @return {@code true} for continuing scheduled update.
+     */
+    protected boolean beforeScheduledStateUpdate(final Instant now,
+                                                 final Service service,
+                                                 final ServiceInstance instance) {
+        return true; // noop
+    }
+
+    /**
+     * Gets the Instant of the last time the state was successfully updated.
+     *
+     * @return the {@link Instant}.
+     */
+    protected Instant lastSucceedStateUpdated() {
+        return lastSucceedStateUpdated;
+    }
+
+
+    /**
+     * Updates the state of the local worker instance.
+     *
+     * @param newState           the new worker state.
+     * @param onStateChangeError the callback to invoke if the state cannot be changed.
+     */
+    protected ServiceInstance updateServiceInstanceState(final Instant now,
+                                                         final Service service,
+                                                         final Service.ServiceState newState,
+                                                         final OnStateTransitionFailureCallback onStateChangeError) {
+
+        // Pre-check the state transition validation with the known local state.
+        ServiceInstance localInstance = localService(service).instance();
+        if (!localInstance.state().isValidTransition(newState)) {
+            log.warn("Failed to transition service [id={}, type={}, hostname={}] from {} to {}. Cause: {}.",
+                localInstance.id(),
+                localInstance.type(),
+                localInstance.server().hostname(),
+                localInstance.state(),
+                newState,
+                "Invalid transition"
+            );
+            mayDisableStateUpdate(service, localInstance);
+            return localInstance;
+        }
+
+        // Ensure only one thread can update any instance at a time.
+        stateLock.lock();
```

Comment:
Ins't this methods caled via a scheduled thread pool each n seconds?
If so, there can't be any concurrent call so the lock is not needed

Reply:
The method can also be called directly from the service. So it's likely to have concurrent access on starting or on shutdown.